packages/middleware/rate-limiter/src/sliding-window.ts

```diff
@@ -0,0 +1,125 @@
+  private readonly samplingPeriodSeconds: number
+  private readonly rateLimit: number
+  private readonly slots: Map<number, Slot> = new Map()
+  private readonly nowSeconds: () => number
+  private readonly slotProvider: SlotProvider
+  private readonly key: string
+  private rateLimitedUntil?: number
+
+  constructor({
+    rate,
+    nowSeconds,
+    slotProvider,
+    key,
+  }: {
+    rate: Rate
+    nowSeconds?: () => number
+    slotProvider: SlotProvider
+    key: string
+  }) {
+    if (rate.period < 1) {
+      throw new Error('Sampling period must be at least 1 second')
+    }
+    if (rate.requests < 0) {
+      throw new Error('Rate limit must be a positive duration')
+    }
+    this.samplingPeriodSeconds = rate.period * (rate.unit === 'hours' ? 3600 : rate.unit === 'minutes' ? 60 : 1)
+    this.rateLimit = rate.requests
+    this.nowSeconds = nowSeconds ?? (() => new Date().getTime() / 1000.0)
+    this.slotProvider = slotProvider
+    this.key = key
+  }
+
+  private getSlots(): [Slot, Slot] {
+    const now = this.nowSeconds()
+    const actualFrom = Math.floor(now / this.samplingPeriodSeconds) * this.samplingPeriodSeconds
+    const oldFrom = actualFrom - this.samplingPeriodSeconds
+
+    const oldSlot =
+      this.slots.get(oldFrom) ??
+      this.slotProvider.create({
+        fromSecond: oldFrom,
+        durationSecond: this.samplingPeriodSeconds,
+        key: this.key,
+      })
+    this.slots.set(oldFrom, oldSlot)
+    const actualSlot =
+      this.slots.get(actualFrom) ??
+      this.slotProvider.create({
+        fromSecond: actualFrom,
+        durationSecond: this.samplingPeriodSeconds,
+        key: this.key,
+      })
+    this.slots.set(actualFrom, actualSlot)
+
+    //Remove expired slots reference
+    if (this.slots.size > 2) {
+      const slots = [...this.slots.values()].sort((slot1, slot2) => slot2.fromSecond - slot1.fromSecond)
+      for (let i = 2; i < slots.length; i++) {
+        this.slots.delete(slots[i].fromSecond)
+      }
+    }
+
+    return [oldSlot, actualSlot]
+  }
+
+  inc(): 'allowed' | 'rate-limited' {
+    const now = this.nowSeconds()
+    if (this.rateLimitedUntil != null && now <= this.rateLimitedUntil) {
+      return 'rate-limited'
+    }
+    this.rateLimitedUntil = undefined
+    const [oldSlot, currentSlot] = this.getSlots()
+    const oldValue = oldSlot.value()
+    const currentValue = currentSlot.value()
+    const oldPart = (this.samplingPeriodSeconds - (now - currentSlot.fromSecond)) / this.samplingPeriodSeconds
+    const requests = currentValue + oldPart * oldValue
+    if (requests < this.rateLimit) {
+      currentSlot.inc()
+      return 'allowed'
+    }
+    this.rateLimitedUntil =
+      oldValue === 0
+        ? currentSlot.fromSecond + this.samplingPeriodSeconds
+        : ((currentValue - this.rateLimit) * this.samplingPeriodSeconds) / oldValue +
+          this.samplingPeriodSeconds +
+          currentSlot.fromSecond
+    return 'rate-limited'
```

Comment:
I'd rather have an explicit else than having to stop for a second to see what happens to the control flow after the if
```suggestion
    } else {
      this.rateLimitedUntil =
        oldValue === 0
          ? currentSlot.fromSecond + this.samplingPeriodSeconds
          : ((currentValue - this.rateLimit) * this.samplingPeriodSeconds) / oldValue +
            this.samplingPeriodSeconds +
            currentSlot.fromSecond
      return 'rate-limited'
    }
```